bika/lims/browser/worksheet/views/analyses.py

```diff
@@ -71,216 +71,361 @@ def __init__(self, context, request):
-                    'before': {},
-                    'after': {},
-                    'choices': {},
-                    'class': {},
-                    'state_class': 'state-empty',
-                    'allow_edit': [],
-                    'colspan': {'Pos':len(self.columns) + len(self.interim_fields)},
-                    'rowspan': {'Pos':1},
-                    'Pos': pos,
-                    'Service': '',
-                    'Attachments': '',
-                    'state_title': 's',})
-                item['replace'] = {
-                    'Pos': "<table width='100%' cellpadding='0' cellspacing='0'>" + \
-                            "<tr><td class='pos'>%s</td>" % pos + \
-                            "<td align='right'>&nbsp;</td></tr></table>",
-                    'select_column': '',
-                    }
-                items.append(item)
-
-        items = sorted(items, key=itemgetter('Pos'))
-
-        slot_items = {} # pos:[item_nrs]
-        for x in range(len(items)):
-            p = items[x]['Pos']
-            if p in slot_items:
-                slot_items[p].append(x)
-            else:
-                slot_items[p] = [x, ]
-        actual_table_position = -1
-        # The first item in items[this position] gets a rowspan for it's
-        # "Position" column, which spans all other table rows in this position.
-        for pos, pos_items in slot_items.items():
-            actual_table_position += 1
-            x = pos_items[0]
-            if pos in empties:
+
+        # Sort the items in accordance with the layout
+        items = sorted(items, key=itemgetter('pos_sortkey'))
+
+        # Fill the slot header cells (first cell of each row). Each slot
+        # contains the analyses that belong to the same parent (AnalysisRequest,
+        # ReferenceSample), so the information about the parent must be
+        # displayed in the first cell of each slot.
+        self.fill_slots_headers(items)
+
+        # Fill empty positions from the layout with fake rows. The worksheet
+        # can be generated by making use of a WorksheetTemplate, so there is the
+        # chance that some slots of this worksheet being empty. We need to
+        # render a row still, at lest to display the slot number (Pos)
+        self.fill_empty_slots(items)
+
+        return items
+
+    def get_uids_strpositions(self):
+        """
+        Returns a dict with the positions of each analysis within the current
+        worksheet in accordance with the current layout. The key of the dict is
+        the uid of the analysis and the value is an string representation of the
+        full position of the analysis within the list:
+            {<analysis_uid>: '<slot_number>:<position_within_slot>',}
+        :return: a dictionary with the full position within the worksheet of
+            all analyses defined in the current layout.
+        """
+        if not self.uids_positions:
```

Comment:
You can skip this cacheing if you do it directly in `__call__` like `self.uids_positions = self.get_uids_strpositions()`